### PR TITLE
Retrofit Whisper backend to use earshot VAD

### LINKED ISSUE #259: Retrofit Whisper backend to use VoiceActivityDetector trait
## Context

The Moonshine integration PR introduced a feature-agnostic \`VoiceActivityDetector\` trait in \`crates/sdr-transcription/src/vad.rs\` along with a sherpa-onnx-backed implementation (\`SherpaSileroVad\`) used by the offline Moonshine session loop.

The trait is deliberately feature-agnostic: it compiles in whisper builds too, with no sherpa-onnx dependency. That was done specifically to make this retrofit straightforward.

## Current Whisper behavior

\`backends/whisper.rs\` uses a naive RMS gate (\`silence_threshold\` slider in the UI) to detect silence between utterances. On RTL-SDR audio, this frequently false-triggers on squelch tails and splits utterances across multiple transcription chunks, which makes the committed text noisy and occasionally cuts words in half.

## Proposed change

1. Add a second \`VoiceActivityDetector\` impl for whisper builds using a pure-Rust Silero crate. Candidates:
   - \`voice_activity_detector\` on crates.io (actively maintained, bundles silero ONNX)
   - \`silero-rs\` (if it exists and is workable)
   - Vendor Silero VAD directly (pull the ONNX file from sherpa-onnx's releases, write a thin inference loop using \`ort\` or \`tract-onnx\`)

2. Wire whisper.rs to use the trait instead of RMS. Retire the \`silence_threshold\` slider (or keep it as a fallback when VAD fails to initialize).

3. Must not add a sherpa-onnx dependency to whisper builds — the feature mutex is non-negotiable.

## Acceptance criteria

- Whisper transcripts stop splitting mid-utterance on squelch tails.
- Whisper builds still compile without any sherpa-onnx deps.
- \`VoiceActivityDetector\` trait shape is unchanged (Moonshine's impl should not need to move).
- Dual-build smoke test (whisper-cuda + sherpa-cpu) still passes.

### PR DESCRIPTION
**Pull request by jasonherald**
## Summary

Fixes #259. Replaces the naive RMS silence gate + fixed 5-second chunking in the Whisper backend with a pure-Rust voice activity detector built on \`earshot\`. Whisper now only decodes bounded speech segments instead of arbitrary 5-second windows that straddle utterance boundaries.

## What was broken

The old flow was:

1. Accumulate 16 kHz mono samples into \`mono_buf\`
2. When \`mono_buf.len() >= 80_000\` (5 seconds), drain a chunk
3. Apply the voice-band spectral gate
4. Gate the chunk by its total RMS — skip if below \`silence_threshold\`
5. Run Whisper inference on the chunk

Problems:

- **Utterance splitting**: a typical NFM transmission is 1–4 s, so two back-to-back transmissions land in different chunks, and a 6+ s transmission gets split mid-word. Whisper's transcript then shows \"and the suspect ran\" / \"toward the building\" as two disconnected fragments.
- **Squelch-tail noise**: the RMS gate is indifferent to frequency content. A noisy squelch tail has enough energy to trip past \`silence_threshold\` and feed hiss straight to Whisper, which hallucinates on it (\"thank you\", \"thanks for watching\", \"subscribe\" — tracked in the \`HALLUCINATIONS\` list this PR leaves untouched).

## What this PR does

Introduces \`backends/earshot_vad.rs\` with \`EarshotVad\`, a
\`VoiceActivityDetector\` implementation that runs a three-state
machine (\`Idle\` / \`Speaking\` / \`HoldingOff\`) over per-frame earshot
voice scores. Each popped segment is a bounded utterance — silence
before, silence after, speech in the middle. Whisper sees nothing
else.

\`backends/whisper.rs\`'s \`run_worker_inner\` is rewritten:

- Accumulated samples go through \`vad.accept()\` instead of a fixed-chunk drain
- Inner loop calls \`vad.pop_segment()\` until empty, passing each segment to a new \`decode_and_emit_segment\` helper
- On channel disconnect, \`vad.flush()\` forces any in-flight segment out so the last utterance isn't dropped

The old \`compute_rms\` helper and its three tests are removed — dead after the gate is gone. \`silence_threshold\` stays in \`BackendConfig\` so the UI slider doesn't break, but whisper.rs no longer reads it. Retiring the slider itself is a follow-up PR.

## Why earshot (not \`voice_activity_detector\`, not Silero directly)

This is the same constraint that caused the whisper/sherpa feature mutex: whisper-rs (bundling whisper.cpp) and any C++ ONNX runtime fight over libstdc++ global state. See #253 for the original heap corruption. Any Silero-based VAD that runs on \`ort\` would have the same class of problem, so we can't use the \`voice_activity_detector\` crate on crates.io (depends on \`ort\`) or wrap Silero directly via \`ort\` / sherpa-onnx.

\`earshot\` is published by pyke (the team that also maintains \`ort\`) explicitly as the pure-Rust alternative:

- **75 KiB embedded neural network** — not Silero, a proprietary smaller model
- **Zero ONNX runtime** — custom mel feature extractor + tiny NN inference loop
- **Only \`libm\` as a dependency** — no C++ state to collide with whisper.cpp
- **Faster than Silero VAD v6** by ~20× per the author's benchmarks (RTF 0.0007 vs Silero's 0.014)
- **MIT/Apache 2.0**, actively maintained

\`cargo tree\` for \`earshot\` is two lines: \`earshot v1.0.0\` → \`libm v0.2.16\`.

## Segmentation tuning

The \`DEFAULT_*\` constants in \`earshot_vad.rs\` mirror sherpa-onnx's Silero defaults so the Whisper and Sherpa backends behave similarly on the same audio:

- \`min_silence_frames\` = 250 ms (15 frames @ 16 ms/frame)
- \`min_speech_frames\` = 250 ms (rejects single-frame noise spikes)
- \`max_speech_frames\` = 20 s (forced flush cap)
- \`voice_threshold\` = 0.5 (earshot's recommended default)

Not exposed to the UI yet — follow-up if real-world testing shows user tuning matters.

## Tests

7 new unit tests in \`earshot_vad::tests\` exercise the pure state machine via direct field manipulation (no recorded audio on disk needed):

- \`fresh_vad_is_empty\` — new instance has nothing to pop
- \`pure_silence_emits_nothing\` — 100 frames of zeros → no segment
- \`partial_frames_across_calls_do_not_drop_samples\` — two half-frame batches end in \`Idle\` with no residual state
- \`reset_clears_all_state\` — \`reset()\` zeros every field
- \`short_burst_is_discarded_as_noise\` — single voiced frame followed by silence → nothing popped
- \`finalize_segment_respects_min_speech_gate\` — below and above the length gate, directly driven
- \`flush_on_active_segment_emits_it\` — forced flush bypasses the trailing-silence requirement

Triple-flavor cargo check + clippy clean on \`whisper-cpu\`, \`whisper-cuda\`, \`sherpa-cpu\`, \`sherpa-cuda\` with \`-D warnings\`.

## Test plan

- [ ] Whisper CUDA build on live NFM with back-to-back transmissions — transcripts should land on utterance boundaries, not every 5 seconds
- [ ] Whisper transcripts should stop splitting mid-word
- [ ] Dead-air chunks should no longer produce hallucinated \"thank you\" / \"subscribe\" text
- [ ] Sherpa builds (cpu and cuda) still compile and work — the feature mutex is respected, earshot is whisper-feature-gated
- [ ] The UI \`silence_threshold\` slider still appears and moves without crashing (backend ignores it, UI is unchanged)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Optional Earshot-based voice activity detector is available as an added VAD backend (enabled with the existing whisper feature).

* **Improvements**
  * Transcription now decodes per detected utterance for more responsive results.
  * In-progress utterances are flushed on disconnect; very short/noise segments are discarded.
  * Replaced fixed-chunk/RMS gating with VAD-driven segmentation.

* **Documentation**
  * Clarified VAD segment semantics and backend-specific padding/holdoff behavior.

* **Tests**
  * Added unit tests covering VAD behaviors and edge cases.

### CODE REVIEW

File: crates/sdr-transcription/src/backends/earshot_vad.rs [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+            State::Idle => {
+                if is_voice {
+                    self.state = State::Speaking;
+                    self.current_segment.clear();
+                    self.speech_frames_in_segment = 0;
+                    self.silence_frames_in_holdoff = 0;
+                    self.current_segment.extend_from_slice(frame);
+                    self.speech_frames_in_segment += 1;
+                }
+                // Idle + silence: discard.
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-transcription/src/backends/earshot_vad.rs]: Great catch — fixed in cb92182 independently, as a result of live scanner testing on this same branch. The user reported "it seems like it's missing the very front of the audio - like it's picking up on the 2nd word", which turned out to be exactly what you describe: earshot's feature extractor uses a 3-frame context window, so the first frame of real speech after silence scores below threshold for lack of history. By the time the score crosses threshold (frame 2 or 3), the leading consonant is gone.

The fix matches your suggestion: a pre-trigger ring buffer of `DEFAULT_LOOKBACK_FRAMES = 4` (64 ms at 16 kHz) held at all times while `Idle`. On `Idle → Speaking` transition, `current_segment.append(&mut self.lookback)` drains the ring into the segment before adding the current frame. `append` empties the source so the ring starts fresh for future Idle phases.

`speech_frames_in_segment` is only incremented for the current (voice-scored) frame, not for lookback frames — lookback is pre-trigger audio and shouldn't inflate the min-speech-length gate.

Validated against live scanner audio after the rebuild: first-word clipping is gone, transmissions land cleanly. Also covered by two new unit tests: `lookback_caps_at_configured_frames_while_idle` (ring stays bounded during long silences) and `lookback_prepends_to_segment_on_transition` (direct field-level simulation of the Idle → Speaking branch verifying segment prefix + voice count).

File: crates/sdr-transcription/src/backends/earshot_vad.rs [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+    }
+
+    fn flush(&mut self) {
+        // Force any in-flight segment to finalize so the caller can
+        // still pop it. Drop any partial frame — it's < 16 ms of
+        // audio, not worth worrying about.
+        self.partial_frame.clear();
+        if !matches!(self.state, State::Idle) {
+            self.finalize_segment();
+        }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-transcription/src/backends/earshot_vad.rs:372]: Fixed in cb92182. Real bug — <16 ms of audio per disconnect isn't huge on a single transmission but on a brisk brevity code it's a meaningful fraction, and it adds up across a session. Refactored `flush()` to:

```rust
fn flush(&mut self) {
    if matches!(self.state, State::Idle) {
        self.partial_frame.clear();
        return;
    }
    if !self.partial_frame.is_empty() {
        self.current_segment.extend_from_slice(&self.partial_frame);
        self.partial_frame.clear();
    }
    self.finalize_segment();
}
```

The Idle branch still drops the partial (there's no segment to attach it to — attaching lookback-adjacent audio to a nonexistent segment would be wrong). Non-Idle branches append the partial to `current_segment` before finalizing so the trailing <16 ms survives.

Covered by two new tests: `flush_preserves_partial_frame_tail_on_active_segment` (stash a half-frame partial with a distinctive pattern, flush, verify the emitted segment ends with the tail pattern) and `flush_drops_partial_frame_when_idle` (the Idle branch still drops, as expected).

File: crates/sdr-transcription/src/backends/earshot_vad.rs [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+//!
+//! # Segmentation model
+//!
+//! The VAD takes 256-sample frames (16 ms at 16 kHz) and returns a
+//! per-frame voice score. This wrapper runs a three-state machine over
+//! those scores to group consecutive voiced frames into segments:
+//!
+//! - **`Idle`** — no speech in flight. Frames below threshold are
+//!   discarded; a frame at or above threshold transitions to
+//!   `Speaking`.
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-transcription/src/backends/earshot_vad.rs]: Fair catch — both had drifted. Fixed in 3195736:

- **Module docstring `Idle` bullet**: rewrote to describe the pre-trigger ring behavior ("retained in a short (`lookback_frames`) pre-trigger ring so the leading consonant of the next utterance isn't clipped") and the drain-on-transition semantics. Added a paragraph explaining that `speech_frames_in_segment` is a running total (not consecutive), that HoldingOff → Speaking bounces still contribute, and that lookback frames intentionally don't count toward it so the gate only measures real voiced content.
- **`speech_frames_in_segment` field comment**: replaced "consecutive voiced frames" with "running total of voice-scored frames across the current segment — NOT consecutive", explaining the bounce-through behavior and the lookback exclusion.

Both were correct at the time of the initial commit but the later lookback work invalidated them.

File: crates/sdr-transcription/src/backends/earshot_vad.rs [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+                } else {
+                    // Idle + silence: append to lookback ring, cap at
+                    // `lookback_frames`. When full, drop the oldest
+                    // frame's worth of samples from the front.
+                    self.lookback.extend_from_slice(frame);
+                    let cap = self.lookback_frames * FRAME_SIZE;
+                    if self.lookback.len() > cap {
+                        let excess = self.lookback.len() - cap;
+                        self.lookback.drain(..excess);
+                    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-transcription/src/backends/earshot_vad.rs]: Fixed in 3195736. Switched `self.lookback` from `Vec<f32>` to `VecDeque<f32>`:

```rust
// Before: O(cap) memmove per frame once the ring is full
self.lookback.extend_from_slice(frame);
let cap = self.lookback_frames * FRAME_SIZE;
if self.lookback.len() > cap {
    let excess = self.lookback.len() - cap;
    self.lookback.drain(..excess);
}

// After: O(excess) head-advancement, no memmove
self.lookback.extend(frame.iter().copied());
let cap = self.lookback_frames * FRAME_SIZE;
while self.lookback.len() > cap {
    self.lookback.pop_front();
}
```

The drain-on-transition path switches from `Vec::append(&mut self.lookback)` to `self.current_segment.extend(self.lookback.drain(..))` which has the same end behavior (lookback empty, all samples moved into `current_segment`). Both existing lookback unit tests (`lookback_caps_at_configured_frames_while_idle` and `lookback_prepends_to_segment_on_transition`) still pass without modification — the first uses `.len()` which is identical between `Vec` and `VecDeque`, and the second directly simulates the transition branch.

Absolute cost matters less here than code quality (60 frames/s × 768-sample memmove was cents on a modern CPU) but the fix is free and it's the right primitive for a sliding window.

File: crates/sdr-transcription/src/backends/earshot_vad.rs [hunk 5]
```diff
@@ -0,0 +1,654 @@
...
+                    // Pause ended — back to Speaking.
+                    self.state = State::Speaking;
+                    self.speech_frames_in_segment += 1;
+                    self.silence_frames_in_holdoff = 0;
+                } else {
+                    self.silence_frames_in_holdoff += 1;
+                    if self.silence_frames_in_holdoff >= self.min_silence_frames {
+                        self.finalize_segment();
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-transcription/src/backends/earshot_vad.rs:277]: Fair catch — the contract was oversold even before `EarshotVad` landed. Fixed in 8c9fb3a by updating the trait docstring in `vad.rs` to describe what implementations actually return rather than what we wished they returned.

Key realizations:

- **`SherpaSileroVad`** was never returning "voiced-only trimmed" segments either — sherpa-onnx wraps Silero's built-in segmenter which applies `speech_pad_ms` (~30 ms default) to both ends of each utterance. So the original contract language was wrong about the existing impl, not just the new one.
- **`EarshotVad`** adds its own 64 ms pre-trigger lookback ring at the start and holds on to any `HoldingOff` trailing-silence frames inside the segment before the `min_silence_frames` gate fires.

The shared invariant is really "a popped segment is a complete utterance boundable by silence, handed off in one piece" — downstream doesn't need to chunk further. Rewrote the docstring around that framing and added concrete per-backend notes for both current impls so future implementers know how much latitude they have with leading/trailing context.

File: crates/sdr-transcription/src/backends/earshot_vad.rs [hunk 6]
```diff
@@ -0,0 +1,654 @@
...
+    #[test]
+    fn short_burst_is_discarded_as_noise() {
+        // A single voiced frame (< min_speech_frames) should be
+        // dropped entirely. Use 1 frame of synthetic voice surrounded
+        // by silence so the segment ends but fails the length gate.
+        let mut vad = EarshotVad::new();
+        vad.accept(&synthetic_voice(1));
+        vad.accept(&silence_frames(DEFAULT_MIN_SILENCE_FRAMES + 5));
+        vad.flush();
+        assert!(vad.pop_segment().is_none());
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-transcription/src/backends/earshot_vad.rs]: Fixed in 8c9fb3a. You're right — `pop_segment().is_none()` was a vacuous assertion because earshot's 3-frame context window makes 1-frame voice inputs unreliable; the VAD could sit in `Idle` the whole test and the assertion would still pass, silently green-lighting a regression in the `min_speech_frames` gate.

Rewrote it to drive the state machine directly, matching the pattern the other state-machine tests already use:

```rust
let mut vad = EarshotVad::new();
vad.state = State::Speaking;
let voiced_count = DEFAULT_MIN_SPEECH_FRAMES - 1;
vad.speech_frames_in_segment = voiced_count;
vad.current_segment
    .extend(std::iter::repeat_n(0.1_f32, voiced_count * FRAME_SIZE));

// Precondition: segment is actually open and sub-gate.
assert_eq!(vad.state, State::Speaking);
assert_eq!(vad.speech_frames_in_segment, voiced_count);
assert!(!vad.current_segment.is_empty());

// Drive finalize via trailing silence.
vad.accept(&silence_frames(DEFAULT_MIN_SILENCE_FRAMES + 5));
vad.flush();

// Three post-conditions pin the discard path.
assert!(vad.pop_segment().is_none());
assert_eq!(vad.state, State::Idle);
assert_eq!(vad.speech_frames_in_segment, 0);
```

The precondition assertions guarantee the VAD was in an open, sub-gate segment before the test exercises finalize, so a future refactor that silently stops seeding can't disguise itself as "nothing emitted → pass". Post-finalize I check three things: no output, state reset to Idle, counters cleared.

File: crates/sdr-transcription/src/backends/earshot_vad.rs [hunk 7]
```diff
@@ -0,0 +1,684 @@
...
+/// transitions on frame 2 or 3 and frame 1 — the actual start of
+/// speech — is lost. 4 frames × 16 ms = 64 ms, which covers the
+/// context gap plus the first consonant/vowel attack.
+const DEFAULT_LOOKBACK_FRAMES: usize = 4;
+
+/// State of the segmentation machine — see module docstring.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum State {
+    /// No segment in progress. Silence frames are discarded.
+    Idle,
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-transcription/src/backends/earshot_vad.rs:122]: Fixed in 6b01b79. That comment had been stale since the lookback work landed — replaced "Silence frames are discarded" with "Incoming frames are retained in `lookback` (the pre-trigger ring, capped at `lookback_frames` × `FRAME_SIZE` samples) so the first voiced frame can pull its leading context along when it transitions to `Speaking`." Keeps the enum variant doc consistent with the other sites that already describe the retention behavior.

File: crates/sdr-transcription/src/backends/earshot_vad.rs [hunk 8]
```diff
@@ -0,0 +1,684 @@
...
+        // update so a max-length segment flushes this frame's audio
+        // before starting a new one.
+        if !matches!(self.state, State::Idle)
+            && self.current_segment.len() >= self.max_speech_frames * FRAME_SIZE
+        {
+            tracing::debug!(
+                frames = self.max_speech_frames,
+                "earshot VAD: max speech frames reached, forcing flush"
+            );
+            self.finalize_segment();
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-transcription/src/backends/earshot_vad.rs:292]: Fixed in 6b01b79. You're right that this branch was under-tested — it's the only thing keeping a stuck mic or a siren from OOMing the process. Added `max_speech_safety_cap_flushes_during_speaking` that drives `process_frame` directly through the safety cap path:

1. Seed VAD into `State::Speaking` with `current_segment` at exactly `max_speech_frames * FRAME_SIZE` samples and `speech_frames_in_segment` well above the min-speech gate (so it'll emit, not discard, on finalize)
2. Assert the preconditions explicitly — segment size, state, speech-frame count, silence counter — so a future refactor that silently breaks the seeding can't quietly green-light the test
3. Call `process_frame` with a silent frame. The match arm appends (pushing segment one frame over cap) and transitions `Speaking → HoldingOff` with silence counter = 1, which is well below `min_silence_frames` so HoldingOff doesn't finalize on its own
4. Assert the full post-conditions: one completed segment, state Idle, both counters cleared, buffer empty

The silent frame is chosen so the cap check is unambiguously the only trigger — a voiced frame would have kept state in Speaking and still exercised the same check, but with a silent frame it's extra clear that HoldingOff didn't self-finalize.
